src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -211,6 +198,54 @@ private async Task ProcessQueueAsync()
             }
         }
 
+        private async Task ProcessSingleWorkItemAsync(QueueItem work, int requestId)
+        {
+            var logIdName = $"{_serverName}.{work.ClientName ?? "Default"}.{requestId++:00000000}.{work.MethodName}.{DateTime.Now:yyyy-MM-dd-HH-mm-ss}";
+            var traceSource = _logger == null ? null : await _logger.CreateTraceSourceAsync(logIdName, _cancelSource.Token).ConfigureAwait(false);
+
+            // Create a linked cancellation token to cancel any requests in progress when this shuts down
+            var cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(_cancelSource.Token, work.CancellationToken).Token;
+
+            var context = CreateRequestContext(work, traceSource);
+
+            if (work.MutatesSolutionState)
+            {
+                // Mutating requests block other requests from starting to ensure an up to date snapshot is used.
+                await ProcessWorkAsync(work, context, traceSource, cancellationToken).ConfigureAwait(false);
+
+                // Now that we've mutated our solution, clear out our saved state to ensure it gets recalculated
+                _lspSolutionCache.Remove(context.Solution.Workspace);
+            }
+            else
+            {
+                // Non mutating are fire-and-forget because they are by definition readonly. Any errors
+                // will be sent back to the client but we can still capture errors in queue processing
+                // via NFW, though these errors don't put us into a bad state as far as the rest of the queue goes.
+                _ = ProcessWorkAsync(work, context, traceSource, cancellationToken);
+            }
+        }
+
+        private static async Task ProcessWorkAsync(QueueItem work, RequestContext context, TraceSource? traceSource, CancellationToken cancellationToken)
+        {
+            try
+            {
+                await work.CallbackAsync(context, cancellationToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                // Ignore cancellation entirely.
+            }
+            catch (Exception e) when (FatalError.ReportAndCatch(e))
+            {
+                // Report all other errors.
```

Comment:
For mutating requests, we need to throw here, in order to ensure the server shuts down, right?

Reply:
Wouldn't it be easier to create the trace source inside the callback here https://github.com/dotnet/roslyn/pull/50416/files#diff-7d61f2773718e7c71639d5423ce0e312f723633b1e5beda588dfe16d4b171f17L134 ?